File: python/tests/test_api.py
"""API-/E2E-Test: echter HTTP-Durchstich durch alle Schichten
(Flask -> Service -> Repository -> MariaDB), im Container gegen
http://localhost:8000."""

import json
import urllib.error
import urllib.request
import uuid

import pytest

import db

BASE = "http://localhost:8000/api/bewerbungen"


def _request(method: str, url: str, body: dict | None = None):
    data = json.dumps(body).encode() if body is not None else None
    req = urllib.request.Request(
        url, data=data, method=method,
        headers={"Content-Type": "application/json"},
    )
    try:
        resp = urllib.request.urlopen(req)
        return resp.status, json.loads(resp.read() or "null")
    except urllib.error.HTTPError as e:
        return e.code, json.loads(e.read() or "null")


@pytest.fixture()
def email():
    addr = f"pyapi+{uuid.uuid4().hex}@example.com"
    yield addr
    # ON DELETE RESTRICT -> erst Bewerbung, dann Bewerber löschen
    conn = db.connect()
    with conn.cursor() as cur:
        cur.execute(
            "DELETE FROM bewerbung WHERE bewerberId IN "
            "(SELECT id FROM bewerber WHERE email = %s)", (addr,))
        cur.execute("DELETE FROM bewerber WHERE email = %s", (addr,))
    conn.close()


def _stelle_id() -> int:
    conn = db.connect()
    with conn.cursor() as cur:
        cur.execute("SELECT MIN(id) AS m FROM stellenangebot")
        row = cur.fetchone()
    conn.close()
    if not row or row["m"] is None:
        pytest.skip("Keine Stelle vorhanden - DB neu initialisieren.")
    return int(row["m"])


def test_post_legt_an_und_get_listet(email):
    status, post = _request("POST", BASE, {
        "vorname": "API", "nachname": "Tester",
        "email": email, "stelle_id": _stelle_id(),
    })
    assert status == 201
    assert "vorgangs_nr" in post

    status, get = _request("GET", BASE)
    assert status == 200
    nummern = [b["vorgangs_nr"] for b in get["bewerbungen"]]
    assert post["vorgangs_nr"] in nummern


def test_post_mit_ungueltigen_daten_400(email):
    status, body = _request("POST", BASE, {"email": "kaputt", "stelle_id": 0})
    assert status == 400
    assert "details" in body

File: ruby/tests/test_api.rb
$LOAD_PATH.unshift File.expand_path("../app", __dir__)

require "minitest/autorun"
require "net/http"
require "uri"
require "json"
require "securerandom"
require "db"

# API-/E2E-Test: echter HTTP-Durchstich gegen den im Container
# laufenden Server (http://localhost:8080).
class BewerbungApiTest < Minitest::Test
  BASE = URI("http://localhost:8080/api/bewerbungen").freeze

  def setup
    @email = "rbapi+#{SecureRandom.hex(8)}@example.com"
  end

  def teardown
    conn = DB.connect
    # ON DELETE RESTRICT -> erst Bewerbung, dann Bewerber loeschen.
    stmt = conn.prepare(
      "DELETE FROM bewerbung WHERE bewerberId IN " \
      "(SELECT id FROM bewerber WHERE email = ?)"
    )
    stmt.execute(@email)
    stmt.close
    stmt = conn.prepare("DELETE FROM bewerber WHERE email = ?")
    stmt.execute(@email)
    stmt.close
    conn.close
  end

  def stelle_id
    conn = DB.connect
    row = conn.query("SELECT MIN(id) AS m FROM stellenangebot").first
    conn.close
    skip "Keine Stelle vorhanden - DB neu initialisieren." if row.nil? || row["m"].nil?
    row["m"].to_i
  end

  def request(method, body = nil)
    http = Net::HTTP.new(BASE.host, BASE.port)
    req = case method
          when "POST" then Net::HTTP::Post.new(BASE.path)
          when "GET"  then Net::HTTP::Get.new(BASE.path)
          end
    req["Content-Type"] = "application/json"
    req.body = JSON.generate(body) if body
    res = http.request(req)
    [res.code.to_i, JSON.parse(res.body || "null")]
  end

  def test_post_legt_an_und_get_listet
    status, post = request("POST",
      vorname: "API", nachname: "Tester",
      email: @email, stelle_id: stelle_id,
    )
    assert_equal 201, status, "Body: #{post.inspect}"
    nummer = post["vorgangs_nr"]
    refute_nil nummer

    status, get = request("GET")
    assert_equal 200, status
    nummern = get["bewerbungen"].map { |b| b["vorgangs_nr"] }
    assert_includes nummern, nummer
  end

  def test_post_mit_ungueltigen_daten_400
    status, body = request("POST", email: "kaputt", stelle_id: 0)
    assert_equal 400, status
    assert body.key?("details"), "details fehlen: #{body.inspect}"
  end
end
